semantic-core/src/Analysis/ScopeGraph.hs

```diff
@@ -1,15 +1,130 @@
+
+instance Monoid ScopeGraph where
+  mempty = ScopeGraph Map.empty
+
+scopeGraph
+  :: Ord term
+  => (forall sig m
+     .  (Carrier sig m, Member (Reader Loc) sig, MonadFail m)
+     => Analysis term User ScopeGraph m
+     -> (term -> m ScopeGraph)
+     -> (term -> m ScopeGraph)
+     )
+  -> [File term]
+  -> (Heap User ScopeGraph, [File (Either (Loc, String) ScopeGraph)])
+scopeGraph eval
+  = run
+  . runFresh
+  . runHeap
+  . traverse (runFile eval)
+
+runFile
+  :: ( Carrier sig m
+     , Effect sig
+     , Member Fresh sig
+     , Member (State (Heap User ScopeGraph)) sig
+     , Ord term
+     )
+  => (forall sig m
+     .  (Carrier sig m, Member (Reader Loc) sig, MonadFail m)
+     => Analysis term User ScopeGraph m
+     -> (term -> m ScopeGraph)
+     -> (term -> m ScopeGraph)
+     )
+  -> File term
+  -> m (File (Either (Loc, String) ScopeGraph))
+runFile eval file = traverse run file
+  where run = runReader (fileLoc file)
+            . runReader (Map.empty @User @Loc)
+            . runFailWithLoc
+            . fmap fold
+            . convergeTerm (Proxy @User) (fix (cacheTerm . eval scopeGraphAnalysis))
+
+-- FIXME: decompose into a product domain and two atomic domains
+scopeGraphAnalysis
+  :: ( Alternative m
+     , Carrier sig m
+     , Member (Reader Loc) sig
+     , Member (Reader (Map.Map User Loc)) sig
+     , Member (State (Heap User ScopeGraph)) sig
+     )
+  => Analysis term User ScopeGraph m
+scopeGraphAnalysis = Analysis{..}
+  where alloc = pure
+        bind name _ m = do
+          loc <- ask @Loc
+          local (Map.insert name loc) m
+        lookupEnv = pure . Just
+        deref addr = do
+          ref <- asks Ref
+          bindLoc <- asks (Map.lookup addr)
+          cell <- gets (Map.lookup addr >=> nonEmpty . Set.toList)
+          maybe (pure Nothing) (foldMapA (pure . Just . mappend (extendBinding addr ref bindLoc))) cell
+        assign addr v = do
+          ref <- asks Ref
+          bindLoc <- asks (Map.lookup addr)
+          modify (Map.insertWith (<>) addr (Set.singleton (extendBinding addr ref bindLoc <> v)))
+        abstract eval name body = do
+          addr <- alloc name
+          assign name (mempty @ScopeGraph)
+          bind name addr (eval body)
+        apply _ f a = pure (f <> a)
+        unit = pure mempty
+        bool _ = pure mempty
+        asBool _ = pure True <|> pure False
+        string _ = pure mempty
+        asString _ = pure mempty
+        record fields = do
+          fields' <- for fields $ \ (k, v) -> do
+            addr <- alloc k
+            loc <- ask @Loc
+            let v' = ScopeGraph (Map.singleton (Decl k loc) mempty) <> v
+            (k, v') <$ assign addr v'
+          pure (foldMap snd fields')
+        _ ... m = pure (Just m)
+
+        extendBinding addr ref bindLoc = ScopeGraph (maybe Map.empty (\ bindLoc -> Map.singleton (Decl addr bindLoc) (Set.singleton ref)) bindLoc)
```

Comment:
`bindLoc` might be `Nothing` if e.g. we’re attempting to read a free variable, or if we’re trying to look a name up outside of its lexical binding scope—indirectly, using records. Thus, record field accesses don’t currently get added to the scope graph, which limits the utility of this approach.